File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.swift
let N = Int(readLine()!)!
var arr:[[Int]] = []
var ch:[[Bool]] = Array(repeating: Array(repeating: false, count: N), count: N)
for _ in 0..<N { arr.append(readLine()!.map{Int(String($0))!}) }

var cnt = 0
var sizes:[Int] = []

var q:[[Int]] = []
let dx = [-1,0,1,0]
let dy = [0,1,0,-1]

for i in 0..<N {
    for j in 0..<N {
        if(arr[i][j] == 1 && !ch[i][j]) {
            cnt += 1
            var size = 1
            q.append([i,j])
            ch[i][j] = true
            while(!q.isEmpty) {
                let cur = q.removeFirst()
                for d in 0..<4 {
                    let nx = cur[0] + dx[d]
                    let ny = cur[1] + dy[d]
                    if(nx > -1 && nx < N && ny > -1 && ny < N && !ch[nx][ny] && arr[nx][ny] == 1) {
                        q.append([nx,ny])
                        ch[nx][ny] = true
                        size += 1
                    }
                }
            }
            sizes.append(size)
        }
    }
}
print(cnt)
sizes.sort()
for size in sizes { print(size) }

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.util.*;
import java.lang.*;
import java.io.*;

class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int N = sc.nextInt();
        int[][] arr = new int[N][N];
        boolean[][] ch = new boolean[N][N];
        
        for(int i=0; i<N; i++) {
            int j = 0;
            char[] input = sc.next().toCharArray();
            for(char c : input) arr[i][j++] = Character.getNumericValue(c);
        }

        int cnt = 0; // 총 단지 수
        List<Integer> sizes = new ArrayList<>(); // 각 단지의 영역 크기
        Queue<int[]> q = new LinkedList<>();

        int[] dx = {-1,0,1,0};
        int[] dy = {0,1,0,-1};

        for(int i=0; i<N; i++) {
            for(int j=0; j<N; j++) {
                if(arr[i][j] == 1 && !ch[i][j]) {
                    q.offer(new int[] {i,j});
                    ch[i][j] = true;
                    cnt++;
                    int size = 1; // 단지 영역 크기
                    while(!q.isEmpty()) {
                        int[] cur = q.poll();
                        for(int d=0; d<4; d++) {
                            int nx = cur[0] + dx[d];
                            int ny = cur[1] + dy[d];
                            if(nx > -1 && nx < N && ny > -1 && ny < N && !ch[nx][ny] && arr[nx][ny] == 1) {
                                q.offer(new int[] {nx,ny});
                                ch[nx][ny] = true;
                                size++;
                            }
                        }
                    }
                    sizes.add(size);
                }
            }
        }

        Collections.sort(sizes);
        System.out.println(cnt);
        for(int size : sizes) System.out.println(size);
    }
}